DESIGN.md

```diff
@@ -0,0 +1,95 @@
+      failure.
+   -  Big subprocesses' start/finish.
+
+## Wanted features
+
+-  _Pull rather than push data from Landonline._ This way we can keep the entire
+   system together, rather than having a costly and awkward split between
+   on-premise and cloud architecture.
+-  _Continuous deployment._ Any time a change is merged to the main branch some
+   smoke tests are performed and the result is deployed to production.
+-  _React to change rather than run on a timer,_ where possible. Ideally the BDE
+   Processor should know about new files as soon as the last file has been
+   closed for writing, and should immediately start processing.
+-  _From-scratch deployment friendly._ That is, we should always have the option
+   to deploy from scratch, deleting _all_ the resources we've created and
+   getting back to an identical state (from the end users' perspective) in a
+   timely fashion (that is, before the next data dump).
+-  _Access for everybody at the org._ This should be simpler than the current
+   solution, and should be easy with read replicas.
+-  _Public access._ This would require a separate user with access to only the
+   appropriately licensed tables (some have a restrictive customer license),
+   using a separate read-only replica.
+-  _Minimal code changes for upstream schema changes._ For example, rather than
+   using complex schema migration code, we should be able to just change the ETL
+   code and redeploy.
+-  Read replica use _does not affect the performance of the pipeline to LDS._
+   Basically just means we'll need to use separate read replicas for them.
+-  Zero configuration. The service should rely on Azure SSO so we don't need to
+   maintain users, we should not need to do any IP whitelisting, and the current
+   version of the BDE Processor should only support the current version of
+   Landonline.
+
+## Non-requirements
+
+-  _Don't keep BDE files around._ They should ideally be kept only in memory or
+   at most in temporary storage which is cleaned up after each run. If we
+   absolutely need to keep them around that should be done as a separate system.
+   If it's convenient, the BDE Processor could read from the data in that system
+   rather than doing its own data copying. TBC who needs the files, and whether
+   _someone_ needs to keep doing this.
+-  _No table versioning._ It creates a _lot_ of extra complexity, and we don't
+   have any PostgreSQL experts at hand to maintain the extensions.
+   -  TBD: Review how versioned tables are currently being used, and work out a
+      reasonable solution.
+
+## Data flow
+
+```mermaid
+graph LR
+   subgraph Landonline
+   informix[(Informix)] -->|daily| bde-repository[bde-repository]
+   end
+
+   subgraph processor-graph[BDE Processor]
+   bde-repository --> processor[/processor/] --> reference[(reference)] --> replica[(read-only replica)]
+   end
+
+   replica --> LDS
+   replica --> public
```

Comment:
To make the database public would still need users to accept a license. CC-BY 4.0 basically. But maybe could have a way to do that?

Reply:
We might be able to add a notice in the documentation, saying that by accessing it you agree to the license agreement. Not sure if that's sufficient, though. I've [asked the Internet](https://opensource.stackexchange.com/q/13595/634); let's see if there's consensus.

Reply:
Yeah. Would need more than that. That's only useful for developers and not other users.

I guess something like the [Basemaps license information on the LINZ website](https://www.linz.govt.nz/products-services/data/licensing-and-using-data/attributing-linz-basemaps-data), coupled with having to go to basemaps.linz.govt.nz to get an API key to kind of ensure people read it.